logprep/connector/s3/output.py

```diff
@@ -327,3 +262,77 @@ def store_failed(self, error_message: str, document_received: dict, document_pro
         }
         self._add_to_backlog(error_document, self._config.error_prefix)
         self._write_to_s3_resource()
+
+    def _add_dates(self, prefix):
+        date_format_matches = self._replace_pattern.findall(prefix)
+        if date_format_matches:
+            now = TimeParser.now()
+            for date_format_match in date_format_matches:
+                formatted_date = now.strftime(date_format_match[2:-1])
+                prefix = re.sub(date_format_match, formatted_date, prefix)
+        return prefix
+
+    @Metric.measure_time()
+    def _write_to_s3_resource(self):
+        """Writes a document into s3 bucket using given prefix."""
+        if self._backlog_size >= self._config.message_backlog_size:
+            self._write_backlog()
+
+    def _add_to_backlog(self, document: dict, prefix: str):
+        """Adds document to backlog and adds a a prefix.
+
+        Parameters
+        ----------
+        document : dict
+           Document to store in backlog.
+        """
+        prefix = self._add_dates(prefix)
+        prefix = f"{self._base_prefix}{prefix}"
+        self._message_backlog[prefix].append(document)
+
+    def _write_backlog(self):
+        """Write to s3 if it is not already writing."""
+        if not self._message_backlog:
+            return
+
+        self._logger.info("Writing %s documents to s3", self._backlog_size)
+        for prefix_mb, document_batch in self._message_backlog.items():
+            self._write_document_batch(document_batch, f"{prefix_mb}/{time()}-{uuid4()}")
+        self._message_backlog.clear()
+
+        if not self._config.call_input_callback:
+            return
+
+        if self.input_connector and hasattr(self.input_connector, "batch_finished_callback"):
+            self.input_connector.batch_finished_callback()
+
+    def _write_document_batch(self, document_batch: dict, identifier: str):
+        try:
+            self._write_to_s3(document_batch, identifier)
+        except EndpointConnectionError as error:
+            raise FatalOutputError(self, "Could not connect to the endpoint URL") from error
+        except ConnectionClosedError as error:
+            raise FatalOutputError(
+                self,
+                "Connection was closed before we received a valid response from endpoint URL",
+            ) from error
+        except (BotoCoreError, ClientError) as error:
+            raise FatalOutputError(self, str(error)) from error
```

Comment:
this try except block is the same than in the setup. Consider to extract it in a decorator to implement a more DRYish way

Reply:
I have added the decorator `_handle_s3_error` that should handle all boto(3) releated exceptions for methods of `S3Output`.